scipy/sparse/linalg/_dsolve/_superluobject.c

```diff
@@ -406,12 +375,29 @@ int NCFormat_from_spMatrix(SuperMatrix * A, int m, int n, int nnz,
             PyErr_SetString(PyExc_TypeError, "Invalid type for array.");
             return -1;
         }
-        Create_CompCol_Matrix(PyArray_TYPE(nzvals),
-                              A, m, n, nnz, PyArray_DATA(nzvals),
-                              (int *) PyArray_DATA(rowind), (int *) PyArray_DATA(colptr),
-                              SLU_NC,
-                              NPY_TYPECODE_TO_SLU(PyArray_TYPE(nzvals)),
-                              SLU_GE);
+        if (csr == 1) {
+            Create_CompRow_Matrix(PyArray_TYPE(nzvals),
+                                  A, m, n, nnz, PyArray_DATA(nzvals),
+                                  (int *) PyArray_DATA(indices), (int *) PyArray_DATA(pointers),
+                                  stype,
+                                  NPY_TYPECODE_TO_SLU(PyArray_TYPE(nzvals)),
+                                  mtype);
+        } else if (csr == 0) {
+            Create_CompCol_Matrix(PyArray_TYPE(nzvals),
+                                  A, m, n, nnz, PyArray_DATA(nzvals),
+                                  (int *) PyArray_DATA(indices), (int *) PyArray_DATA(pointers),
+                                  stype,
+                                  NPY_TYPECODE_TO_SLU(PyArray_TYPE(nzvals)),
+                                  mtype);
+        } else if (csr == -1) {
+            Create_SuperNode_Matrix(PyArray_TYPE(nzvals),
+                                    A, m, n, nnz, PyArray_DATA(nzvals),
+                                    (int *) PyArray_DATA(pointers), (int *) PyArray_DATA(indices),
+                                    (int *) PyArray_DATA(pointers), identity_col_to_sup, identity_sup_to_col,
```

Comment:
This looks strange, but I think it is correct. According to [documentation in the SuperLU code](https://github.com/scipy/scipy/blob/1ecd8dfba6c1484d0c5f252c4d79df6d7f8a2c49/scipy/sparse/linalg/_dsolve/SuperLU/SRC/supermatrix.h#L96-L118), a supernodal CSC matrix stores
- the non-zero values in a contiguous array `nzval`, like for a normal CSC matrix,
- the row indices of all the non-zero values in an array `rowind`; however, unlike normal CSC, for consecutive columns of the same sparsity structure, the row indices do not need to be repeated;
- accordingly, instead of a single `colptr` (as in normal CSC) there are now two arrays `rowind_colptr` and `nzval_colptr` storing, at index j, the position in `rowind` and `nzval` where the data for column j starts;
- and some more information to translate between indices of columns and indices of groups of columns with the same sparsity structure.

We have to use the supernodal format because that is what `dgstrs` expects as the type of the argument. However, we have no information about whether there are consecutive columns with the same sparsity structure. This is why in our case, `rowind_colptr` and `nzval_colptr` have the same entries from `pointers`.

As above, it's not documented by SuperLU (to my knowledge) whether the two arrays may be the same portion of memory in case they have the same content... but it seems to work.

Reply:
Thanks for the detailed answer! SGTM.